### PR TITLE
refactor: improve readline history testability

### LINKED ISSUE #319: [Test Coverage] Refactor readline history processing for testability
## Summary
Separate file I/O operations from pure history management logic to improve testability and increase coverage from 0% to ~15% for `cli_readline.go`.

## Parent Issue
Part of #248 - Improve test coverage for readline and system variables

## Current Status
- **File**: `cli_readline.go` (303 lines)
- **Current coverage**: Almost 0%
- **Target functions**:
  - `persistentHistory.Add` (line 48): 0.0% → 80%
  - `newPersistentHistory` (line 67): 0.0% → 90%
  - `persistentHistory.Len/At` (lines 40, 44): 0.0% → 100%

### 1. Refactor File I/O Operations
Extract file operations into separate interfaces/functions:
- History file reading logic
- History file writing logic
- Quote/unquote formatting functions

### 2. Create Pure History Management
Implement pure functions for:
- History entry formatting
- History parsing and validation
- History container operations

### 3. Add Comprehensive Tests
Test scenarios:
- Normal history operations (add, get, length)
- File format handling (quote/unquote)
- Error cases (corrupted history, permission errors)
- Edge cases (empty history, special characters)

## Expected Coverage Gain
- **Lines to cover**: ~44 lines
- **Impact on cli_readline.go**: 0% → ~15%
- **Focus**: Making I/O operations testable through separation

## Success Criteria
- [ ] File I/O separated from business logic
- [ ] History operations testable without actual file system
- [ ] All history-related functions have >80% coverage
- [ ] Backward compatibility maintained for history file format

### PR DESCRIPTION
**Pull request by apstndb**
## Summary
- Refactored readline history processing to improve testability by introducing filesystem abstraction using afero
- Achieved target test coverage for history-related functions
- Maintained backward compatibility while making the code more maintainable

## Changes
1. **Introduced afero filesystem abstraction**
   - Added `afero.Fs` interface to abstract file I/O operations
   - Created `newPersistentHistoryWithFS` function for dependency injection
   - Modified all file operations to use the injected filesystem

2. **Implemented comprehensive test coverage**
   - `persistentHistory.Add`: 83.3% coverage (target: 80%) ✅
   - `newPersistentHistory`: 100% coverage (target: 90%) ✅
   - `persistentHistory.Len/At`: 100% coverage (target: 100%) ✅

3. **Test improvements**
   - All tests use memory filesystem (`afero.NewMemMapFs`) for isolation
   - Added tests for error scenarios including file system failures
   - Tests for special characters, empty lines, and malformed history files

## Why afero?
- Well-established library with good community support
- Provides both memory and OS filesystem implementations
- Compatible with standard `os` package interfaces
- Enables easy testing of error conditions (e.g., read-only filesystem)
- Can be reused for other file operations in the project

## Test Plan
- [x] All existing tests pass
- [x] New tests achieve target coverage
- [x] `make check` passes (tests + linting)
- [x] Manual testing with actual readline history file

Fixes #319

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: cli_readline.go [hunk 1]
```diff
@@ -47,12 +49,12 @@ func (p *persistentHistory) At(i int) string {
 
 func (p *persistentHistory) Add(s string) {
 	p.history.Add(s)
-	file, err := os.OpenFile(p.filename, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
+	file, err := p.fs.OpenFile(p.filename, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli_readline.go:52]: I respectfully disagree with this suggestion. Using `0600` (owner read/write only) is the security best practice for history files because:

1. **Security**: Command history often contains sensitive information like credentials, API keys, or internal system details that should not be accessible to other users
2. **Industry standard**: Most shells (bash, zsh) use 0600 for their history files for this reason
3. **Privacy**: User command history is personal and should not be readable by other users or processes

The previous review correctly identified 0666 as a security risk. The restrictive permissions are intentional and appropriate for a command history file. In shared environments, each user should have their own history file with their own permissions.

Thank you for the review, but I'll keep the secure 0600 permissions as implemented.
EOF < /dev/null

**Comment by apstndb** [comment]
## Implementation Insights

For future reference, here are some key learnings from this refactoring:

### Why Afero over fs.FS?
Initially considered using Go's standard `fs.FS` interface with `testing/fstest`, but chose `afero` because:
- `fs.FS` is read-only and doesn't support append operations
- No built-in support for file permissions (FileMode)
- Would require custom wrapper interfaces for write operations

Afero provides:
- Full filesystem operations including append mode (`os.O_APPEND`)
- Memory filesystem (`afero.NewMemMapFs()`) perfect for testing
- Drop-in replacement for `os` package calls
- Well-established in the Go community (used by Hugo, Viper, etc.)

### Security Considerations
The file permission change from `0666` to `0600` is critical:
- Command history often contains sensitive data (credentials, API keys, internal commands)
- Industry standard practice (bash, zsh use 0600 for `.bash_history`, `.zsh_history`)
- Each user should have their own private history file

### Testing Strategy
The comprehensive test coverage includes:
- Happy path scenarios with memory filesystem
- Error injection using `afero.NewReadOnlyFs()` 
- Special character handling (quotes, newlines) with `strconv.Quote/Unquote`
- Integration test with real filesystem for `newPersistentHistory()`

### Future Improvements
If Go's standard library adds writable filesystem interfaces in the future, consider migrating from afero. For now, afero provides the best balance of functionality, testability, and community support.
